temporal-sdk/src/main/java/io/temporal/internal/worker/LocalActivityWorker.java

```diff
@@ -260,8 +262,56 @@ public boolean dispatch(
           new LocalActivityExecutionContext(
               params, resultCallback, localRetryDeadline, scheduleToCloseDeadline);
 
-      LocalActivityAttemptTask task =
-          new LocalActivityAttemptTask(executionContext, params.getInitialActivityTask());
+      PollActivityTaskQueueResponse activityTask = params.getInitialActivityTask();
+      LocalActivityAttemptTask task = new LocalActivityAttemptTask(executionContext, activityTask);
+
+      final Failure previousExecutionFailure = executionContext.getPreviousExecutionFailure();
+      if (previousExecutionFailure != null) {
+        // This is not an original local execution, it's a continuation using a workflow timer.
+        // We should verify if the RetryOptions currently supplied in the workflow still allow the
+        // retry.
+        // If not, we need to recreate the same structure of an error like it would happen before we
+        // started to sleep on the timer, at the end of the previous local execution.
+        RetryDecision retryDecision =
+            shouldRetry(executionContext, task.getAttemptTask(), null, previousExecutionFailure);
+        if (!retryDecision.doNextAttempt()) {
+          Failure failure;
+          if (RetryState.RETRY_STATE_TIMEOUT.equals(retryDecision.retryState)) {
+            if (previousExecutionFailure.hasTimeoutFailureInfo()
+                && TimeoutType.TIMEOUT_TYPE_START_TO_CLOSE.equals(
+                    previousExecutionFailure.getTimeoutFailureInfo().getTimeoutType())) {
+              // This scenario should behave the same way as a startToClose timeout happening and
+              // encountering
+              // RetryState#TIMEOUT during calculation of the next attempt (which is effectively a
+              // scheduleToClose
+              // timeout).
+              // See how StartToCloseTimeoutHandler or
+              // io.temporal.internal.testservice.StateMachines#timeoutActivityTask
+              // discard startToClose in this case and replaces it with scheduleToClose
+              failure =
+                  newTimeoutFailure(
+                      TimeoutType.TIMEOUT_TYPE_SCHEDULE_TO_CLOSE,
```

Comment:
Weird. I wouldn't expect this to suddenly become `SCHEDULE_TO_CLOSE`

I would expect the failure to be whatever the reason would be because of the current retry policy. EX: if they changed the max attempts to be fewer than were already done, it should be max attempts reached. If they shortened some timeout, it should be timeout failure of that type, etc.

Reply:
Weird. But this is how Temporal Server works right now.
See this compliance test:
https://github.com/temporalio/sdk-java/blob/c5f0ebbe8a5cd52ee018809ecdd466379b5a7339/temporal-sdk/src/test/java/io/temporal/workflow/activityTests/ActivityTimeoutTest.java#L325
And this piece in Test Server that I wrote to comply with the real server behavior:
https://github.com/temporalio/sdk-java/blob/7b526740eeff0d0e3568716281d0aaf86a344bc8/temporal-test-server/src/main/java/io/temporal/internal/testservice/StateMachines.java#L1614